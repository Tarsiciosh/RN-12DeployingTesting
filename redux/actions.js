import { login } from '../api'

// action types
export const UPDATE_CONTACT = 'UPDATE_CONTACT'
export const UPDATE_USER = 'UDPATE_USER'
export const LOG_IN_SENT = 'LOG_IN_SENT'
export const LOG_IN_FULFILLED = 'LOG_IN_FULFILLED'
export const LOG_IN_REJECTED = 'LOG_IN_REJECTED'

// actions creators
export const updateUser = update => ({
  type: UPDATE_USER,
  payload: update,
})

export const addContact = newContact => ({
  type: UPDATE_CONTACT,
  payload: newContact
})

// async action creator                      // DEPENDENCY INJECTION 
export const logInUser = (username, password, loginFn = login) => async dispatch => {
  // "=" default function values 
  //const realLoginFn = loginFn ? loginFn : login
  //const realLoginFn = loginFn || login

  dispatch({ type: LOG_IN_SENT }) // here the argument to dispatch is an obj
  try {
    const token = await /*realLoginFn*/ loginFn(username, password)
    dispatch({ type: LOG_IN_FULFILLED, payload: token })
  } catch (err) {
    dispatch({ type: LOG_IN_REJECTED, payload: err.message })
  }
}

// this fucntion logInUser takes a username and a password as parameters and it
// return an async function that takes a function as a parameter called dispatch and return void
// it has to be an async function because inside it calls an async function
