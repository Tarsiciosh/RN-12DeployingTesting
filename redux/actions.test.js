import * as actions from './actions'
// import everything as an abject called actions
/* global test, expect */

/*
test('updateUser returns an action', () => {
  expect(actions.updateUser({name: 'test name'})).toStrictEqual({
    type: actions.UPDATE_USER, 
    payload: {name: 'test name'}, 
  })
  expect(actions.updateUser({phone: '1234567890'})).toStrictEqual({
    type: actions.UPDATE_USER,
    payload: {phone: '1234567890'}
  })
})
*/

/*
   test('updateUser returns an action', () => {
    expect(actions.updateUser({name: 'test name'})).toMatchSnapshot()
  })
*/

describe('updateUser returns actions', () => {
  it('returns an action', () => {
    expect(actions.updateUser({name: 'test name'})).toMatchSnapshot()
  })

  it('handles an empty object', () => {
    expect(actions.updateUser({})).toMatchSnapshot()
  })
    
  it('handles an empty name', () => {
    expect(actions.updateUser({name:''})).toMatchSnapshot()
  })
}) 


describe('logInUser returns action', () => {
  //const dispatch = () => {}
  const errMessage = 'incorrect credentials'
  const fakeToken = 'thisIsATestToken' 
  const mockLogin = (username, password) => {
    if (username === 'u' && password === 'p') {
      return fakeToken
    } else {
      throw new Error(errMessage)
    }
  }

  it('dispatches LOG_IN_SENT', async () => {
    const mockDispatch = jest.fn()
    await actions.logInUser('','')(mockDispatch)
    //all the args that the mock fn was invoked on
    console.log(mockDispatch.mock.calls)
    expect(mockDispatch.mock.calls[0][0]).toEqual({type: 'LOG_IN_SENT'})
  })

  it('dispatches LOG_IN_FULFILLED with correct credentials', async () => {
    const mockDispatch = jest.fn()
    await actions.logInUser('u', 'p', mockLogin)(mockDispatch)  
    expect(mockDispatch.mock.calls[1][0]).toEqual({type: actions.LOG_IN_FULFILLED, payload: fakeToken})
    expect(mockDispatch.mock.calls[1]).toMatchSnapshot()
  })

  it('dispatches LOG_IN_REJECTED with incorrect credentials', async () => {
    const mockDispatch = jest.fn()
    await actions.logInUser('', '', mockLogin)(mockDispatch)  
    expect(mockDispatch.mock.calls[1][0]).toEqual({type: actions.LOG_IN_REJECTED, payload: errMessage})
    expect(mockDispatch.mock.calls[1]).toMatchSnapshot()
  })

})

mockLogin = (username, password) =>
{
  return {token: 'this is a real token'}
}