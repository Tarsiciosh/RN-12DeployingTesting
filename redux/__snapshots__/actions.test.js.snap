// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`logInUser returns action dispatches LOG_IN_FULFILLED with correct credentials 1`] = `
Array [
  Object {
    "payload": "thisIsATestToken",
    "type": "LOG_IN_FULFILLED",
  },
]
`;

exports[`logInUser returns action dispatches LOG_IN_REJECTED with incorrect credentials 1`] = `
Array [
  Object {
    "payload": "incorrect credentials",
    "type": "LOG_IN_REJECTED",
  },
]
`;

exports[`updateUser returns actions handles an empty name 1`] = `
Object {
  "payload": Object {
    "name": "",
  },
  "type": "UDPATE_USER",
}
`;

exports[`updateUser returns actions handles an empty object 1`] = `
Object {
  "payload": Object {},
  "type": "UDPATE_USER",
}
`;

exports[`updateUser returns actions returns an action 1`] = `
Object {
  "payload": Object {
    "name": "test name",
  },
  "type": "UDPATE_USER",
}
`;
